### PR TITLE
Adjust coupon user limit logic

### LINKED ISSUE #16422: Coupon Usage Limit Not Working as Expected
Regarding Coupons- I'm finding that the _Usage Limits > Usage Limit Per User_ setting is not functioning as expected. Whenever setting the usage limit per user to 1, for example, I would expect that the coupon would only be able to be used once as detailed (logged in users would be checked by User ID and guests by billing email address.)

However, upon testing this numerous times, I'm seeing very inconsistent results. Some emails I tested allowed me to use the coupon twice, another email three times, and another email only once (as expected.) All times, I tested this in incognito mode, as a guest, and I received drastically different results each time. Since I was testing as a guest, it should've been checking my billing email address used, but it's using something else instead it seems. 

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create a Coupon with the following settings:

Coupon Settings General

Coupon Settings Usage Restriction

Coupon Settings Usage Limits

2. Place an order using this coupon with any email of your choice. The order should go through successfully the first time:

Order Received

3. Place a second order using this coupon with a different email address. At this point, I received a _Coupon usage limit has been reached_ error, even though I was using a completely new billing email address:

Coupon usage limit has been reached

4. Place a third & forth order using this coupon but this time with the same email address. For me, I was able to place an order twice using the same email address. Sometimes, up to three times. Other times (using new emails to test), I got rejected as expected. Continue to test various email addresses in this fashion until you see the behavior. 

## Expected/actual behavior

I would expect that as a logged out user, it would use my billing email address to cross-check how many times that coupon was used, but it seems it is not successful in doing so:

Usage Limit based on billing email

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://latestupdate.mystagingwebsite.com
Site URL: https://latestupdate.mystagingwebsite.com
WC Version: 3.1.1
Log Directory Writable: ✔
WP Version: 4.8.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: nginx
PHP Version: 7.0.21
PHP Post Max Size: 100 MB
PHP Time Limit: 300
PHP Max Input Vars: 6144
cURL Version: 7.54.1
OpenSSL/1.0.1t

SUHOSIN Installed: –
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.1.1
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.1.1

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/

### Theme ###

Name: Storefront
Version: 2.2.4
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

### PR DESCRIPTION
**Pull request by mikejolley**
Fixes #16422

When logged out, compares the given billing email against user and billing emails of existing users to find matches.

### CODE REVIEW

**Comment by claudiulodro** [comment]
Looks good. Works well.

**Comment by hehafner** [comment]
Hi, I'm getting the same issue. But more specifically, it seems to do it in Chrome. This is happening to customers who have never purchased products. We tested on Firefox and did not experience this issue. I don't do a lot of coding, can you dumb it down a little and explain what I need to do to fix this? I would greatly appreciate any help.

**Comment by cbubny** [comment]
After updating to woocommerce 3.2.1, we too are experiencing this issue.  The usage limits field stopped working for all coupons. Any value aside from unlimited usage produces a Coupon has reached it's limit notice to the end user.  After receiving notifications from customers attempting to checkout with the coupons, we tested it ourselves. We verified we were not logged in, and tried new transactions with email addresses that have never been used.  We get the error every time.

**Comment by keeneyemedia** [comment]
I can confirm we're seeing the same issue as @cbubny on our end. _Any_ limit gives a "Coupon usage limit has been reached." error.

**Comment by Kyzcreig** [comment]
We're seeing the same issue in 3.2.1

**Comment by mikejolley** [comment]
Just tested a new coupon with per-user limit and it applied ok both logged in and as guest.

**Comment by Kyzcreig** [comment]
Did you discover the issue? Or did it resolve with the latest update of Woo? Maybe we should clear our cache?

**Comment by glagonikas** [comment]
@mikejolley this appears to be an issue with WCv3.2 and beyond, as it didn't happen on 3.1.

The logic to check if the coupon is used is probably incorrect. See below for the explanation of the issue

## Steps to reproduce the issue
1. Disable all plugins apart from WooCommerce and WooCommerce Stripe
2. Create a new coupon and set "Usage limit per user" to 1
3. Make a purchase of a product using the newly created coupon as a new customer (Customer A)
4. Clear your session and attempt to make another purchase as a new customer (Customer B)

## Expected behavior
Allow both users to complete the transaction and purchase the product

## Actual behavior
An error stating "Coupon usage limit has been reached." is showing on the checkout for Customer B, preventing them from placing the order

## Possible cause
https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart.php#L1494-L1502
should probably be changed to
`
$users_query = new WP_User_Query( array(
    'fields'       => 'ID',
    'meta_query'   => array(
    		array(
        	'key'      => '_billing_email',
        	'value'    => $check_emails,
        	'compare'  => 'IN',
        )
    ),
) );
`

**Comment by mattblancarte** [comment]
@glagonikas I can confirm that meta_query nested array fixes the issue on my builds.

The login in WC_Discounts::is_coupon_valid is fine and that behaves as expected.

The bug occurs on the `woocommerce_after_checkout_validation` hook, which calls WC_Cart::check_customer_coupons

**Comment by mikejolley** [comment]
Seen https://github.com/woocommerce/woocommerce/pull/17237?

**Comment by glagonikas** [comment]
Thanks for making this fix into a pull request and merging.

@mikejolley what's the roadmap for 3.2.2?

**Comment by OzzDizz** [comment]
We are now experiencing this issue in 3.2.1 with ALL coupon codes site-wide that have a usage limit (per user) being rejected for new or first-time users. In the shopping cart they can add the coupon code and the discount will be calculated, but then when they proceed to the checkout they get an error message that the usage limit has been reached - even if they've never used the coupon code before. We've tested this with new coupons, new user accounts, logged in, not logged in, and get the same results as described above: any usage limit causes the coupon code to not work, and our customers are reminding us several times a day!

**Comment by glagonikas** [comment]
@OzzDizz, this bug first appeared on v3.2.0 and will be fixed on 3.2.2.

As a quick fix, you can either downgrade to v3.1 or just replace the code in the woocommerce plugin with the one i posted few comments above. (/wp-content/plugins/woocommerce/includes/class-wc-cart.php)

**Comment by nanoanno** [comment]
I'm having the same issue on 3.2.1 ... hopefully we will have an update soon :)
